### ISSUE TITLE
The spinoffs: HyperApp 🎢 

### ISSUE THREAD

**Issue by nichoth** [hyperapp/hyperapp#68]
Just fyi for anyone interested, the funness of this repo inspired me to make something similar. The differences being that the view rendering is decoupled, so you can plug in any render function, and it's all es5 — no build tools required. Thanks @jbucaran for making a cool thing.

https://github.com/nichoth/h-app

**Comment by jbucaran** [comment]
@nichoth By the way, HyperApp is also all ES5, no build tools required. We're using rollup to generate the UMD bundles that users grab when embedding via an `<script>` tag, but that's another thing.

**Comment by nichoth** [comment]
Oh! I didn't realize that :) that's important

**Comment by nichoth** [comment]
The es6 was a false alarm :). The thing I would really like to solve is finding an easy way to run automated tests on browser applications. I'm not an expert with elm — I don't know how tests work — but compile time type checking would catch a majority of the bugs I have day to day. JS is in a weird place where unit tests often fill the role of type checking. My interest in decoupling model & view is because I'm trying to find a way to easily write automated app tests. One idea is to declare all the events as data (the `.msg` object), so we can easily test the model with mock input against the events. I'm just kind of thinking out loud here, would be interested in hearing what other's experiences are like.

**Comment by jbucaran** [comment]
Do you mean HyperApp tests or your app's tests?

**Comment by nichoth** [comment]
@jbucaran I mean my app's tests (tests of the application, not the dependencies). Dependencies are easier because the interface is not UI via the dom.

**Comment by jbucaran** [comment]
> My interest in decoupling model & view

ELI5, what do you mean by decoupling? I see in your `Counter` example you've split the app into counter.js and app.js. 

You can do the same in HyperApp from day one.

app.js
const Counter = require("./counter")

**Comment by nichoth** [comment]
It is decoupled because we provide defaults that can be overridden. The render function applied before the view state. So the default is

```js
require('h-app')(someModel)
```

But the files are factored in such a way that you can do
```js
var App = require('h-app/app')
var app = App(mySpecialRenderer)

app(myModel)

function mySpecialRenderer (tree) {
    // ...
    // this gets the return value from myModel.view
}
```

This necessitates an observable interface for the app state (via `observ-struct` in my lib), so the renderer can subscribe to changes, whereas in hyperapp the render function is baked in.

**Comment by jbucaran** [comment]
What you want to decouple the `render` function, correct?

**Comment by nichoth** [comment]
Yes, basically. The end goal is being able to write tests, the hypothesis is that we can expose the app state without using the Dom.

**Comment by jbucaran** [comment]
> @nichoth The end goal is being able to write tests

This implies you can't write tests now, but that's not true.

---

I'm also considering separating the render function in order support stuff like Electron, kinda like it went with Snabbdom and I think I can split things up and still stay under 1kb, since we're using rollup.

**Comment by nichoth** [comment]
@jbucaran what would your application tests look like in hyperapp?

**Comment by jbucaran** [comment]
Would probably use jest like we are with HyperApp.

I'm juggling a lot of things at once, so I don't have a ready repo to share with you. Perhaps you could contribute one?

**Comment by nichoth** [comment]
yes It's one of my main goals at the moment :) . Where I'm at is that writing tests is 100% possible, but often too laborious to be worthwhile/feasible, and also not obvious as to what/how to test. Part of this varies depending on what kind of experiences we are coming from with applications and teams of developers. 

I do have some beef with frameworks like jest, in that it seems too complicated, and I don't want to be dependent on something that *needs* a large corporation maintaining it because it is so vast. Ideally I would like to not use any framework that takes more than a weekend to read and understand.

**Comment by jbucaran** [comment]
I was unsure about jest at first, for pretty much the same reasons, but after being introduced to it by @maraisr, the more I use it, the more enjoyable I find it. I also wanted to try something new and the tooling is minimal.

In the future, when we make render document agnostic, I may switch back to tape. I'm happy with it so far, so this tape or jest, that's not a priority at the moment.

**Comment by jbucaran** [comment]
Hi @nichoth! How's h-app coming along? 👋 

After a month, I think we can close here, since I don't see how is this an issue about hyperapp.

**Comment by nichoth** [comment]
Yeah sounds good. I have been experimenting more with different APIs. I made a traditional todomvc with pull-streams, which I like. https://github.com/nichoth/todomvc-pull-stream . The important part is [pull-stream-model](https://github.com/nichoth/pull-stream-model), which provides some structure/convention.

So main thing was to find a nice way of writing tests, which I still haven't gotten around to, but will soon. In this case we are adding tests to an existing large app that doesn't have any.

The idea is that all the app events we care about are serialized to static objects that can be logged. So in the todo app, https://github.com/nichoth/todomvc-pull-stream/blob/master/app.js#L33 , you can log events at any point. In the example it's just the browser console, but in real life we have a small utility that sends events to a websocket server that logs them. Then you can replay the log in an automated way and check that the events are valid and the resulting state is ok.

**Comment by jbucaran** [comment]
Sounds like mad science :) 👍
